text/0000-process-stdio-redirection.md

```diff
@@ -0,0 +1,126 @@
+would allow redirecting stdio with a physical file or even another process (via
+OS pipe) without forcing the parent process to buffer the data itself.
+
+For example, one may wish to spawn a process which prints gigabytes
+of data (e.g. logs) and use another process to filter through it, and save the
+result to a file. The current API would force the parent process to dedicate a
+thread just to buffer all data between child processes and disk, which is
+impractical given that the OS can stream the data for us via pipes and file
+redirection.
+
+# Detailed design
+
+First, the standard library should provide an OS agnostic way of creating OS
+in-memory pipes, i.e. a `Pipe`, providing reader and writer handles as
+`PipeReader` and `PipeWriter` respectively. The reader and writer will simply be
+an abstraction over the OS specific file descriptor/HANDLE, and will implement
+the respective `Read`/`Write` trait, making it impossible to confuse the two
+together. In addition, each should implement the appropriate `AsRaw{Fd,
+Handle}`/`FromRaw{Fd, Handle}` for wrapping/unwrapping of the OS handles.
+
+On Unix systems the pipe can be created with `libc::pipe` and the file
+descriptors wrapped as appropriate (e.g. `sys::fs2::File`). On Windows the pipe
+can be created via Windows' `CreatePipe` (a stub for which is missing in
+`liblibc` at the moment) and the resulting HANDLEs also appropriately wrapped
+(using `sys::fs2::File`).
+
+This proposal considers making the `Pipe`'s reader and writer fields public
+to allow easily moving ownership of the two handles, but any other appropriate
+interface is acceptable.
+
+```rust
+pub struct Pipe {
+	pub reader: PipeReader,
+	pub writer: PipeWriter,
+}
+
+pub struct PipeReader(sys::fs2::File);
+pub struct PipeWriter(sys::fs2::File);
+
+impl Read for PipeReader { ... }
+impl Write for PipeWriter { ... }
+```
+
+Next, several `redirect_*` methods should be added to `Stdio` for certain
+"blessed" types offered by the standard library, such as `File`, `PipeRead`, and
+`PipeWrite`. By white-listing the accepted file-like types we can ensure the API
+and its behavior is consistent across Windows and Unix.
+
+```rust
+fn redirect_file(f: File) -> Stdio { ... }
+fn redirect_pipe_read(r: PipeRead) -> Stdio { ... }
+fn redirect_pipe_write(w: PipeWrite) -> Stdio { ... }
+```
+
+These methods should take ownership of their arguments since storing references
+will require `Stdio` and `Command` to gain lifetime parameters, which will break
+the currently stabilized implementations. Thus the caller will be responsible
+for duplicating their handles appropriately if they wish to retain ownership.
+
+To make redirections easier to use `Stdio` should become clonable so that once a
+file-like handle is wrapped, the wrapper can be passed to any number of
+`Commands` simultaneously. This can be accomplished by reference counting the
+wrapped file handle.
+
+```rust
+impl Clone for Stdio { ... }
```

Comment:
I was thinking something along the lines of a `StdioExt` trait which will be defined in `std::os::$patform::io` based on the appropriate `AsRawFoo`, but their method signatures would match.

```rust
// Unix
pub trait StdioExt { fn redirect<T: AsRawFd>(t: &T) -> Stdio { ... } }
// Windows
pub trait StdioExt { fn redirect<T: AsRawHandle>(t: &T) -> Stdio { ... } }
```

So then you could write something like the following in both Unix and Windows
```rust
#[cfg(unix)] use std::os::unix::io::StdioExt as StdioExt;
#[cfg(windows)] use std::os::windows::io::StdioExt as StdioExt;

let file = File::open(...);
let stdio = unsafe { StdioExt::redirect(file) };
...
```

Instead of having to write the following everywhere you want to redirect something
```rust
#[cfg(unix)] use std::os::unix::io::StdioExt as StdioExt;
#[cfg(windows)] use std::os::windows::io::StdioExt as StdioExt;

let file = File::open(...);
let stdio = unsafe {
    if cfg!(unix) {
        StdioExt::from_raw_fd(file)
    } else {
        StdioExt::from_raw_handle(file)
    }
};
...
```

This way users are still somewhat aware that this is platform specific code, without suffering the constant need of `cfg`.